packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,253 @@ import { Far } from '@endo/marshal';
+  //       const helloAddress = await E(host).publish(hello);
+  //       // helloAddress = '/alleged-chain/${chainID}/egress/${clistIndex}'
+  //       return handler;
+  //     },
+  //   });
+  //
+  // B:
+  //
+  //   const { host, handler } = await E(agoric.vattp).makeNetworkHost('ag-chain-A', comms);
+  //   E(agoric.ibcport[1]).connect('/ibc-port/portADDR/ordered/vattp-1', handler);
+  //   E(E(host).lookup(helloAddress)).hello();
+
+  const nextConnectionNumberBaggageKey = 'networkHostNextConnectionNumber';
+  let nextConnectionNumber = provide(
+    baggage,
+    nextConnectionNumberBaggageKey,
+    () => 7n,
+  );
+  // TODO: Should this collection be Weak (non-iterable)?
+  const networkHostNames = provideDurableSetStore(baggage, 'networkHostNames');
+  const networkHostKindHandle = provideKindHandle(baggage, 'networkHost');
+  const initNetworkHostState = (allegedName, comms, console) => {
+    assert(
+      !networkHostNames.has(allegedName),
+      X`already have host for ${allegedName}`,
+    );
+    networkHostNames.add(allegedName);
+    const connectionName = `connection-${nextConnectionNumber}`;
+    nextConnectionNumber += 1n;
+    baggage.set(nextConnectionNumberBaggageKey, nextConnectionNumber);
+    return {
+      comms,
+      console,
+
+      allegedName,
+      connection: null,
+      connectionName,
+      exportedObjects: makeScalarBigMapStore(
+        `network host exported objects by address: ${connectionName} as ${allegedName}`,
+        { durable: true },
+      ),
+      nextAddressNumber: 1n,
+      // This assumes one connection per host.
+      // If that is not the case, receiver should instead be a
+      // durable makeScalarBigWeakMapStore keyed by connection.
+      receiver: null,
+      remoteLocatorUnum: null,
+    };
+  };
+  const makeNetworkHost = defineDurableKindMulti(
+    networkHostKindHandle,
+    initNetworkHostState,
+    {
+      host: {
+        publish: ({ state }, obj) => {
+          const address = `/alleged-name/${state.allegedName}/egress/${state.nextAddressNumber}`;
+          state.nextAddressNumber += 1n;
+          state.exportedObjects.init(address, obj);
+          return address;
+        },
+        lookup: ({ state }, address) => {
+          return E(state.remoteLocatorUnum).lookup(address);
+        },
+      },
+      commsSetReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.receiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.receiver = receiver;
+        },
+      },
+      commsTransmitter: {
+        transmit: ({ state }, msg) => {
+          // 'msg' will be a string (vats/comms/outbound.js deliverToRemote)
+          E(state.connection).send(msg);
+        },
+      },
+      localLocatorUnum: {
+        lookup: ({ state }, address) => {
+          return state.exportedObjects.get(address);
+        },
+      },
```

Comment:
Objects like these are needed for the comms vat handshaking in `onOpen`, and making them internal-use facets of the network host multi-kind is the best durability approach I could think of. But I'm open to other options.

Reply:
that sounds good to me